TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -0,0 +1,435 @@
+			     _resolution,
+			     float(0.) );
+
+}
+
+
+
+void TOFEstimators::init() { 
+
+  streamlog_out(DEBUG) << "   init called  " << std::endl ;
+
+  // usually a good idea to
+  printParameters() ;
+
+  _nRun = 0 ;
+  _nEvt = 0 ;
+
+  // initialize gsl random generator
+  _rng = gsl_rng_alloc(gsl_rng_ranlxs2);
+
+  Global::EVENTSEEDER->registerProcessor(this);
+
+
+  _TOFNames = { "TOFFirstHit",
+		"TOFClosestHits", "TOFClosestHitsError",
+		"TOFCluster",     "TOFClusterError" } ;
+
+}
+
+
+void TOFEstimators::processRunHeader( LCRunHeader* ) {
+
+  _nRun++ ;
+} 
+
+
+
+void TOFEstimators::processEvent( LCEvent * evt ) { 
+
+
+  // use the global Marlin random seed for this processor
+  gsl_rng_set( _rng, Global::EVENTSEEDER->getSeed(this) ) ;   
+
+  streamlog_out(DEBUG ) << "seed set to " << Global::EVENTSEEDER->getSeed(this) << std::endl;
+
+  streamlog_out(DEBUG2) << "   process event: " << evt->getEventNumber() 
+			<< "   in run:  " << evt->getRunNumber() << std::endl ;
+
+
+  // get the PFO collection from the event if it exists
+  LCCollection* colPFO = nullptr ;
+    
+  try{
+    colPFO = evt->getCollection( _colNamePFO ) ;
+  }
+  catch(lcio::Exception){
+    streamlog_out( DEBUG6 ) << " collection " << _colNamePFO
+			    << " not found in event - nothing to do ... " << std::endl ;
+  }
+
+  if( colPFO->getTypeName() != LCIO::RECONSTRUCTEDPARTICLE ) {
+
+    streamlog_out( ERROR ) << " collection " << _colNamePFO
+			   << " not of type LCIO::RECONSTRUCTEDPARTICLE " << std::endl ;
+
+    colPFO = nullptr ;
+  }
+    
+  if( colPFO != nullptr ){
+    
+
+
+    PIDHandler pidh( colPFO );
+    int algoID = pidh.addAlgorithm( name()  , _TOFNames);
+
+
+    int nPFO = colPFO->getNumberOfElements()  ;
+    
+    
+    for(int i=0; i< nPFO ; ++i){ 
+
+      ReconstructedParticle* pfo = dynamic_cast<ReconstructedParticle*>(  colPFO->getElementAt( i ) ) ;
+
+      bool isCharged = false ;
+
+      if(  pfo->getClusters().size() != 1 ){
+	
+	streamlog_out( DEBUG1 ) << " ignore particle w/ cluster number other than one:  " <<  *pfo << std::endl ; 
+	continue ;
+      }
+
+      if( std::fabs( pfo->getCharge() ) < 0.1  && pfo->getTracks().size() == 0  ) {
+	
+	isCharged = false ;
+      }
+      
+      else if ( std::fabs( pfo->getCharge() ) > 0.1  && pfo->getTracks().size() == 1  ) {
+	
+	isCharged = true ;
+	
+      } else {
+	
+	streamlog_out( DEBUG1 ) << " ignore particle w/ track number other than zero or one:  " <<  *pfo << std::endl ; 
+	continue ;
+      }
+
+      streamlog_out( DEBUG1 ) << " --- compute TOF estimators for particle : " << *pfo << std::endl ;
+
+
+      
+      // =======  use only Ecal hits  (requires the CalorimeterHitType to be set in the digitizer )
+      //          with time information ( > 1 ps) and layer <= max layer
+
+      int maxLayerNum = _maxLayerNum ;
+      std::function<bool(CalorimeterHit*)> selectHits =  [maxLayerNum](CalorimeterHit* h){
+
+	return ( isEcal( h )            &&
```

Comment:
Why do you use only ecal hits ? What about neutral hadrons ?

Reply:
At this point it is a feature that we restrict ourselves to the Ecal only. Combining hits from Ecal and Hcal with different time resolutions would probably require a more involved algorithm.
We can leave this for a future version.